x-pack/plugins/fleet/server/services/agents/current_upgrades.ts

```diff
@@ -0,0 +1,150 @@
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core/server';
+import pMap from 'p-map';
+
+import type { FleetServerAgentAction, CurrentUpgrade } from '../../types';
+import { AGENT_ACTIONS_INDEX, AGENT_ACTIONS_RESULTS_INDEX } from '../../../common';
+import { SO_SEARCH_LIMIT } from '../../constants';
+
+/**
+ * Return current bulk upgrades (non completed or cancelled)
+ */
+export async function getCurrentBulkUpgrades(
+  esClient: ElasticsearchClient,
+  now = new Date().toISOString()
+): Promise<CurrentUpgrade[]> {
+  // Fetch all non expired actions
+  const [_upgradeActions, cancelledActionIds] = await Promise.all([
+    _getUpgradeActions(esClient, now),
+    _getCancelledActionId(esClient, now),
+  ]);
+
+  let upgradeActions = _upgradeActions.filter(
+    (action) => cancelledActionIds.indexOf(action.actionId) < 0
+  );
+
+  // Fetch acknowledged result for every upgrade action
+  upgradeActions = await pMap(
+    upgradeActions,
+    async (upgradeAction) => {
+      const { count } = await esClient.count({
+        index: AGENT_ACTIONS_RESULTS_INDEX,
+        ignore_unavailable: true,
+        query: {
+          bool: {
+            must: [
+              {
+                term: {
+                  action_id: upgradeAction.actionId,
+                },
+              },
+            ],
+          },
+        },
+      });
+
+      return {
+        ...upgradeAction,
+        nbAgentsAck: count,
+        complete: upgradeAction.nbAgents <= count,
+      };
+    },
+    { concurrency: 20 }
+  );
+
+  upgradeActions = upgradeActions.filter((action) => !action.complete);
+
+  return upgradeActions;
+}
+
+async function _getCancelledActionId(
+  esClient: ElasticsearchClient,
+  now = new Date().toISOString()
+) {
+  const res = await esClient.search<FleetServerAgentAction>({
+    index: AGENT_ACTIONS_INDEX,
+    ignore_unavailable: true,
+    size: SO_SEARCH_LIMIT,
+    query: {
+      bool: {
+        must: [
+          {
+            term: {
+              type: 'CANCEL',
+            },
+          },
+          {
+            exists: {
+              field: 'agents',
+            },
+          },
+          {
+            range: {
+              expiration: { gte: now },
+            },
+          },
+        ],
+      },
+    },
+  });
+
+  return res.hits.hits.map((hit) => hit._source?.data?.target_id as string);
+}
+
+async function _getUpgradeActions(esClient: ElasticsearchClient, now = new Date().toISOString()) {
+  const res = await esClient.search<FleetServerAgentAction>({
+    index: AGENT_ACTIONS_INDEX,
+    ignore_unavailable: true,
+    size: SO_SEARCH_LIMIT,
+    query: {
+      bool: {
+        must: [
+          {
+            term: {
+              type: 'UPGRADE',
+            },
+          },
+          {
+            exists: {
+              field: 'agents',
+            },
+          },
+          {
+            range: {
+              expiration: { gte: now },
```

Comment:
Note: I moved this code out of `upgrade.ts` which already has the logic of upgrade action.

Should we remove the expired condition from here? So in case the upgrade takes longer, we would still see the callout (current expiration for immediate upgrades is set to 2h).

Reply:
I think removing the expiration condition from this query makes sense.

Reply:
I am changing this part in the next pr for Agent activity flyout, so will leave as is for this pr.